### PR TITLE
Adding support for metering

### PR DESCRIPTION
**Pull request by vijasrin**
- Outbound metering P4 model changes in BMV2
- Changes to support the SAI api generation for metering tables
- Skipping failing SAIC tests till the metadata has been updated

### CODE REVIEW

**Comment by vijasrin** [comment]
@prsunny

File: dash-pipeline/bmv2/dash_pipeline.p4 [hunk 1]
```diff
@@ -231,6 +235,98 @@ control dash_ingress(
...
+            meta.eni_id: exact @name("meta.eni_id:eni_id");
+        }
+        actions = { NoAction; }
+#ifdef TARGET_BMV2_V1MODEL
+        counters = meter_bucket_outbound_counter;
+#endif
+    }
+
+#ifdef TARGET_BMV2_V1MODEL
+    direct_counter(CounterType.bytes) meter_bucket_inbound_counter;
```
Comment (ashutosh-agrawal) [inline dash-pipeline/bmv2/dash_pipeline.p4:315]: For the inbound case, does the meter_class need to be added to the inbound_routing table actions?

Comment (vijasrin) [reply dash-pipeline/bmv2/dash_pipeline.p4:315]: yes. will address inbound meter in a separate PR

**Comment by chrispsommers** [comment]
Some meeting comments:
- explore indirect counters (Andy)
- optional attributes shouldn't have to be provided in test-cases; code-generator is asserting on number of attributes; needs fix?

**Comment by vijasrin** [comment]
SAI API for ref: https://github.com/opencomputeproject/SAI/compare/master...vijasrin:SAI:meter_sai?diff=unified

File: dash-pipeline/bmv2/dash_pipeline.p4 [hunk 2]
```diff
@@ -231,6 +235,98 @@ control dash_ingress(
...
+        } else {
+            if (meta.is_overlay_ip_v6 == 0) {
+                meta.dropped = true;
+            }
+        }
+    }
+
+    @name("meter_policy|dash_meter")
+    @Sai[isobject="true"]
+    table meter_policy {
```
Comment (svshah-intel) [inline dash-pipeline/bmv2/dash_pipeline.p4:252]: I did not see requirement in HLD for what this table is trying to achieve? Is it needed here as part of metering? What this table is doing mostly is dropping cross address family packets.. Vendors may be able to achieve it in a different way. Is it required as part of metering policy? If so, may be we should update requirement section in HLD

File: dash-pipeline/bmv2/dash_pipeline.p4 [hunk 3]
```diff
@@ -231,6 +235,98 @@ control dash_ingress(
...
+            @Sai[type="sai_uint32_t", isreadonly="true"] bit<32> inbound_bytes_counter) {
+        // read only counters for SAI api generation
+    }
+
+    @name("meter_bucket|dash_meter")
+    @Sai[isobject="true"]
+    table meter_bucket {
+        key = {
+            meta.eni_id: exact @name("meta.eni_id:eni_id");
+            meta.meter_class: exact @name("meta.meter_class:meter_class");
```
Comment (svshah-intel) [inline dash-pipeline/bmv2/dash_pipeline.p4:291]: what is not clear is if meter_class across all the types globally unique or each policy, route, mapping have their own scope of meter_class values? It is very important to be specified some-where. I did not see it mentioned in HLD. If scope of bit<16> meter_class is unique within scope of each policy, route, mapping separately then, have we not lost the context here while incrementing count here that will be reported.

File: dash-pipeline/bmv2/dash_pipeline.p4 (hunk 2 above)
Comment (vijasrin) [reply dash-pipeline/bmv2/dash_pipeline.p4:252]: This table corresponds to the DASH_METER_POLICY table mentioned in the HLD. It is used to create v4 and v6 meter policy table instances which are then associated with ENIs. This is similar to the ACL_GROUP table for ipv4 and ipv6 address families. It is not necessary for vendor implementations to introduce this table in their datapath as long as behavior is the same

File: dash-pipeline/bmv2/dash_pipeline.p4 [hunk 4]
```diff
@@ -231,6 +235,98 @@ control dash_ingress(
...
+    }
+
+    @name("meter_policy|dash_meter")
+    @Sai[isobject="true"]
+    table meter_policy {
+        key = {
+            meta.meter_policy_id : exact @name("meta.meter_policy_id:meter_policy_id");
+        }
+        actions = {
+            set_meter_policy_attrs();
```
[the remark this answers is not in the document]
Comment (vijasrin) [reply dash-pipeline/bmv2/dash_pipeline.p4:257]: ok. changing to check_ip_addr_family

File: dash-pipeline/bmv2/dash_pipeline.p4 [hunk 5]
```diff
@@ -231,6 +235,98 @@ control dash_ingress(
...
+    @name("meter_rule|dash_meter")
+    @Sai[isobject="true"]
+    table meter_rule {
+        key = {
+            meta.meter_policy_id: exact @name("meta.meter_policy_id:meter_policy_id") @Sai[type="sai_object_id_t", isresourcetype="true", objects="METER_POLICY"];
+            hdr.ipv4.dst_addr : ternary @name("hdr.ipv4.dst_addr:dip");
+        }
+
+     actions = {
+            meter_rule_action;
```
[the remark this answers is not in the document]
Comment (vijasrin) [reply dash-pipeline/bmv2/dash_pipeline.p4:274]: ok. updated name

File: dash-pipeline/bmv2/dash_pipeline.p4 (hunk 3 above)
Comment (vijasrin) [reply dash-pipeline/bmv2/dash_pipeline.p4:291]: The scope of meter class-id is global and is managed above the SAI layer. The meter class-id usage and examples are mentioned in the HLD

File: dash-pipeline/SAI/sai_api_gen.py [hunk 6]
```diff
@@ -54,13 +54,26 @@ def p4_annotation_to_sai_attr(p4rt, sai_attr):
...
                     exit(1)
     sai_attr['field'] = sai_type_to_field[sai_attr['type']]
 
+def p4_annotation_to_sai_table(p4rt, sai_table):
+    for anno in p4rt[STRUCTURED_ANNOTATIONS_TAG]:
+        if anno[NAME_TAG] == SAI_TAG:
+            for kv in anno[KV_PAIR_LIST_TAG][KV_PAIRS_TAG]:
+                if kv['key'] == 'isobject':
+                    sai_table['is_object'] = kv['value']['stringValue']
+                if kv['key'] == 'ignoretable':
```
[the remark this answers is not in the document]
Comment (vijasrin) [reply dash-pipeline/SAI/sai_api_gen.py:74]: will do. thnx!

File: dash-pipeline/bmv2/dash_pipeline.p4 [hunk 7]
```diff
@@ -231,6 +235,98 @@ control dash_ingress(
...
+            @defaultonly deny;
+        }
+        const default_action = deny;
```
[the remark this answers is not in the document]
Comment (vijasrin) [reply dash-pipeline/bmv2/dash_pipeline.p4]: yes, updated

**Comment by vijasrin** [comment]
Adding skip for the SAIC tests till the metadata is updated

**Comment by taras-keryk-plv** [comment]
CI fails for two reasons:
1. metadata is missing in SAI-C because the headers are generated automatically.
Here is visible what information is missing: https://github.com/taras-keryk-plv/SAI-Challenger/pull/81/files

2. dpugen does not generate config for these new attributes. Therefore, the tests will fail even if #1 is fixed.
So we have the next errors:

saiserver: saidasheni.cpp:376: sai_status_t sai_create_eni(sai_object_id_t*, sai_object_id_t, uint32_t, const sai_attribute_t*): Assertion `(matchedParams == expectedParams)' failed.

[413296.894310] traps: saiserver[643376] general protection fault ip:7f5b56bdd941 sp:7f5b4f7fd420 error:0 in libc-2.31.so[7f5b56bdd000+178000]

**Comment by chrispsommers** [comment]
Hi @taras-keryk-plv could you be more specific which test is failing where dpugen is involved? We can explore updating it. Thank you!- @mgheorghe please take note

**Comment by taras-keryk-plv** [comment]
Hi @chrispsommers, @mgheorghe 
yes, i tried:
```
make run-saichallenger-client-bash
cd functional
./run-tests.sh --setup=sai_dpu_client_server_snappi.json tutorial/test_sai_vnet_outbound_small_scale_config_via_dpugen.py::TestSaiVnetOutbound
```
Thank you!

**Comment by chrispsommers** [comment]
Hi @vijasrin PR https://github.com/sonic-net/DASH/pull/378 was merged, it should address item #1 in https://github.com/sonic-net/DASH/pull/365#issuecomment-1559030778. Assuming this is true can you merge from main, re-enable the tests you'd skipped (except for any still depending upon dpugen enhancements)? Thanks.

File: dash-pipeline/bmv2/dash_metadata.p4 [hunk 8]
```diff
@@ -62,6 +62,14 @@ struct metadata_t {
     bit<16> stage3_dash_acl_group_id;
     bit<16> stage4_dash_acl_group_id;
     bit<16> stage5_dash_acl_group_id;
+    bit<1> meter_policy_en;
+    bit<1> mapping_meter_class_override;
+    bit<16> meter_policy_id;
+    bit<16> policy_meter_class;
+    bit<16> route_meter_class;
+    bit<16> mapping_meter_class;
```
Comment (svshah-intel) [inline dash-pipeline/bmv2/dash_metadata.p4:70]: is your PR's scope to update route and mapping meter classes via some actions? If so I did not see them in the PR

File: dash-pipeline/bmv2/dash_outbound.p4 [hunk 9]
```diff
@@ -9,19 +9,32 @@
 control outbound(inout headers_t hdr,
                  inout metadata_t meta)
 {
-    action route_vnet(bit<16> dst_vnet_id) {
+    action set_route_meter_attrs(bit<1> meter_policy_en,
+                                 bit<16> meter_class) {
+        meta.meter_policy_en = meter_policy_en;
+        meta.meter_class = meter_class;
```
Comment (svshah-intel) [inline dash-pipeline/bmv2/dash_outbound.p4:15]: did you mean to assign meter_class here or route_meter_class? and then later use it in the condition below where you pick policy/route/mapping meter_class?

Comment (vijasrin) [reply dash-pipeline/bmv2/dash_outbound.p4:15]: yes, that is right. fixed this. thnx!

File: dash-pipeline/bmv2/dash_metadata.p4 (hunk 8 above)
Comment (vijasrin) [reply dash-pipeline/bmv2/dash_metadata.p4:70]: fixed. thnx!

**Comment by chrispsommers** [comment]
@reshmaintel mentioned there might be some forthcoming test cases for new outbound metering. Will confirm.

File: dash-pipeline/bmv2/dash_pipeline.p4 [hunk 10]
```diff
@@ -388,6 +462,33 @@ control dash_ingress(
...
+            meter_rule.apply();
+        }
+
+        {
+            if (meta.meter_policy_en == 1) {
+                meta.meter_class = meta.policy_meter_class;
+            } else {
+                meta.meter_class = meta.route_meter_class;
+            }
+            if ((meta.meter_class == 0) || (meta.mapping_meter_class_override == 1)) {
```
[the remark this answers is not in the document]
Comment (vijasrin) [reply dash-pipeline/bmv2/dash_pipeline.p4:476]: thnx for pointing this out! will sync up on this and update the HLD in a separate PR @prsunny

[the remark this answers is not in the document]
Comment (svshah-intel) [reply dash-pipeline/bmv2/dash_pipeline.p4:476]: please update once confirmed if change in HLD necessitates change in your PR or remains unchanged. Once confirmed, your PR remains unchanged (and change is only in HLD), I don't have any further comments

[the remark this answers is not in the document]
Comment (vijasrin) [reply dash-pipeline/bmv2/dash_pipeline.p4:476]: @prsunny to update the metering requirements section in the HLD to reflect the current logic in bmv2. No further changes are required to the p4 code. @svshah-intel plz approve if no further comments. Thnx!

[the remark this answers is not in the document]
Comment (prsunny) [reply dash-pipeline/bmv2/dash_pipeline.p4:476]: Ack

File: dash-pipeline/bmv2/dash_pipeline.p4 [hunk 11]
```diff
@@ -231,6 +235,76 @@ control dash_ingress(
...
+    table meter_policy {
+        key = {
+            meta.meter_policy_id : exact @name("meta.meter_policy_id:meter_policy_id");
+        }
+        actions = {
+            check_ip_addr_family;
+        }
+    }
+
+    action set_meter_class(bit<16> meter_class) {
```
Comment (svshah-intel) [inline dash-pipeline/bmv2/dash_pipeline.p4:261]: re-name this action to "set_policy_meter_class" ?

**Review by svshah-intel — APPROVED** [review]

**Pull request merged**